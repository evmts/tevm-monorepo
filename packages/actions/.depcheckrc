ignores: [
  "@ethereumjs/evm",
  "@ethereumjs/block",
  "@ethereumjs/blockchain",
  "@ethereumjs/vm",
  "@tevm/actions-types",
  "@tevm/errors"
]
skip-missing: true
