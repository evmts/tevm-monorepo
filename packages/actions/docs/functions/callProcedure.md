[**@tevm/actions**](../README.md)

***

[@tevm/actions](../globals.md) / callProcedure

# Function: callProcedure()

> **callProcedure**(`client`): [`CallJsonRpcProcedure`](../type-aliases/CallJsonRpcProcedure.md)

Defined in: [packages/actions/src/Call/callProcedure.js:10](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/Call/callProcedure.js#L10)

Creates a Call JSON-RPC Procedure for handling call requests with Ethereumjs EVM

## Parameters

### client

`TevmNode`\<`"fork"` \| `"normal"`, \{ \}\>

## Returns

[`CallJsonRpcProcedure`](../type-aliases/CallJsonRpcProcedure.md)
