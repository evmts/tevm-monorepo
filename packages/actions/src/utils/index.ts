export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
