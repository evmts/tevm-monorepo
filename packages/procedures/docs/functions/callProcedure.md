**@tevm/procedures** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > callProcedure

# Function: callProcedure()

> **callProcedure**(`client`): `CallJsonRpcProcedure`

Creates a Call JSON-RPC Procedure for handling call requests with Ethereumjs EVM

## Parameters

▪ **client**: `object`

▪ **client.extend**: \<`TExtension`\>(`decorator`) => `BaseClient`\<`"fork"` \| `"proxy"` \| `"normal"`, `object` & `TExtension`\>

Extends the base client with additional functionality. This enables optimal code splitting
and extensibility

▪ **client.forkUrl?**: `string`

Fork url if the EVM is forked

**Example**

```ts
const client = createMemoryClient({ forkUrl: 'https://mainnet.infura.io/v3/your-api-key' })
console.log(client.forkUrl)
```

▪ **client.getChainId**: () => `Promise`\<`number`\>

Gets the chainId of the current EVM

**Example**

```ts
const client = createMemoryClient()
const chainId = await client.getChainId()
console.log(chainId)
```

▪ **client.getVm**: () => `Promise`\<`TevmVm`\>

Internal instance of the VM. Can be used for lower level operations.
Normally not recomended to use unless building libraries or extensions
on top of Tevm.

▪ **client.mode**: `"fork"` \| `"proxy"` \| `"normal"`

The mode the current client is running in
`fork` mode will fetch and cache all state from the block forked from the provided URL
`proxy` mode will fetch all state from the latest block of the provided proxy URL
`normal` mode will not fetch any state and will only run the EVM in memory

**Example**

```ts
let client = createMemoryClient()
console.log(client.mode) // 'normal'
client = createMemoryClient({ forkUrl: 'https://mainnet.infura.io/v3/your-api-key' })
console.log(client.mode) // 'fork'
```

▪ **client.ready**: () => `Promise`\<`true`\>

Returns promise that resulves when the client is ready
The client is usable without calling this method but may
have extra latency on the first call from initialization

**Example**

```ts
const client = createMemoryClient()
await client.ready()
```

▪ **client.setChainId**: (`chainId`) => `void`

Sets the chain id of the current EVM

## Returns

## Source

[procedures/src/tevm/callProcedure.js:9](https://github.com/evmts/tevm-monorepo/blob/main/packages/procedures/src/tevm/callProcedure.js#L9)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
