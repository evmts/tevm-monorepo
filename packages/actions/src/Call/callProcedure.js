import { hexToBigInt, numberToHex } from '@tevm/utils'
import { parseBlockTag } from '../utils/parseBlockTag.js'
import { callHandler } from './callHandler.js'

/**
 * Creates a Call JSON-RPC Procedure for handling call requests with Ethereumjs EVM
 * @param {import('@tevm/node').TevmNode} client
 * @returns {import('./CallJsonRpcProcedure.js').CallJsonRpcProcedure}
 */
export const callProcedure = (client) => async (request) => {
	const { errors = [], ...result } = await callHandler(client)({
		throwOnFail: false,
		...(request.params[1]
			? {
					stateOverrideSet: Object.fromEntries(
						Object.entries(request.params[1]).map(([address, state]) => [
							address,
							{
								...(state.code ? { code: state.code } : {}),
								...(state.balance ? { balance: hexToBigInt(state.balance) } : {}),
								...(state.nonce ? { nonce: hexToBigInt(state.nonce) } : {}),
								...(state.state ? { state: state.state } : {}),
								...(state.stateDiff ? { stateDiff: state.stateDiff } : {}),
							},
						]),
					),
				}
			: {}),
		...(request.params[2]
			? {
					blockOverrideSet: {
						...(request.params[2].blobBaseFee ? { blobBaseFee: hexToBigInt(request.params[2].blobBaseFee) } : {}),
						...(request.params[2].baseFee ? { baseFee: hexToBigInt(request.params[2].baseFee) } : {}),
						...(request.params[2].gasLimit ? { gasLimit: hexToBigInt(request.params[2].gasLimit) } : {}),
						...(request.params[2].coinbase ? { coinbase: request.params[2].coinbase } : {}),
						...(request.params[2].time ? { time: hexToBigInt(request.params[2].time) } : {}),
						...(request.params[2].number ? { number: hexToBigInt(request.params[2].number) } : {}),
					},
				}
			: {}),
		...(request.params[0].code ? { code: request.params[0].code } : {}),
		...(request.params[0].data ? { data: request.params[0].data } : {}),
		...(request.params[0].deployedBytecode ? { deployedBytecode: request.params[0].deployedBytecode } : {}),
		...(request.params[0].createTrace ? { createTrace: request.params[0].createTrace } : {}),
		...(request.params[0].createAccessList ? { createAccessList: request.params[0].createAccessList } : {}),
		...(request.params[0].blobVersionedHashes ? { blobVersionedHashes: request.params[0].blobVersionedHashes } : {}),
		...(request.params[0].caller ? { caller: request.params[0].caller } : {}),
		...(request.params[0].depth ? { depth: request.params[0].depth } : {}),
		...(request.params[0].gasPrice ? { gasPrice: hexToBigInt(request.params[0].gasPrice) } : {}),
		...(request.params[0].gas ? { gas: hexToBigInt(request.params[0].gas) } : {}),
		...(request.params[0].gasRefund ? { gasRefund: hexToBigInt(request.params[0].gasRefund) } : {}),
		...(request.params[0].origin ? { origin: request.params[0].origin } : {}),
		...(request.params[0].salt ? { salt: request.params[0].salt } : {}),
		...(request.params[0].selfdestruct ? { selfdestruct: new Set(request.params[0].selfdestruct) } : {}),
		...(request.params[0].skipBalance ? { skipBalance: request.params[0].skipBalance } : {}),
		...(request.params[0].to ? { to: request.params[0].to } : {}),
		...(request.params[0].value ? { value: hexToBigInt(request.params[0].value) } : {}),
		...(request.params[0].blockTag ? { blockTag: parseBlockTag(request.params[0].blockTag) } : {}),
		...(request.params[0].createTransaction ? { createTransaction: request.params[0].createTransaction } : {}),
		...(request.params[0].from ? { from: request.params[0].from } : {}),
		...(request.params[0].maxFeePerGas ? { maxFeePerGas: hexToBigInt(request.params[0].maxFeePerGas) } : {}),
		...(request.params[0].maxPriorityFeePerGas
			? { maxPriorityFeePerGas: hexToBigInt(request.params[0].maxPriorityFeePerGas) }
			: {}),
	})
	if (errors.length > 0) {
		const error = /** @type {import('./TevmCallError.js').TevmCallError}*/ (errors[0])
		return {
			jsonrpc: '2.0',
			error: {
				code: error.code,
				message: error.message,
				data: {
					errors: errors.map(({ message }) => message),
				},
			},
			method: 'tevm_call',
			...(request.id === undefined ? {} : { id: request.id }),
		}
	}

	/**
	 * @type {Record<`0x${string}`, Set<import('@tevm/utils').Hex>> | undefined}
	 */
	const accessList =
		result.accessList !== undefined
			? Object.fromEntries(Object.entries(result.accessList).map(([key, value]) => [key, value]))
			: undefined

	/**
	 * @type {import('./CallJsonRpcResponse.js').CallJsonRpcResponse}
	 */
	const out = {
		jsonrpc: '2.0',
		result: {
			executionGasUsed: result.executionGasUsed,
			rawData: result.rawData,
			...(result.selfdestruct ? { selfdestruct: result.selfdestruct } : {}),
			...(result.gasRefund ? { gasRefund: result.gasRefund } : {}),
			...(result.gas ? { gas: result.gas } : {}),
			...(result.logs ? { logs: result.logs } : {}),
			...(result.txHash ? { txHash: result.txHash } : {}),
			...(result.blobGasUsed ? { blobGasUsed: result.blobGasUsed } : {}),
			...(accessList !== undefined ? { accessList } : {}),
			...(result.preimages ? { preimages: result.preimages } : {}),
			...(result.l1Fee ? { l1Fee: result.l1Fee } : {}),
			...(result.l1BaseFee ? { l1BaseFee: result.l1BaseFee } : {}),
			...(result.l1BlobFee ? { l1BlobFee: result.l1BlobFee } : {}),
			...(result.l1GasUsed ? { l1GasUsed: result.l1GasUsed } : {}),
			...(result.amountSpent ? { amountSpent: result.amountSpent } : {}),
			...(result.baseFee ? { baseFee: result.baseFee } : {}),
			...(result.totalGasSpent ? { totalGasSpent: result.totalGasSpent } : {}),
			...(result.trace
				? {
						trace: {
							...result.trace,
							gas: result.trace.gas,
							structLogs: result.trace.structLogs.map((log) => ({
								...log,
								gas: log.gas,
								gasCost: log.gasCost,
								stack: log.stack,
							})),
						},
					}
				: {}),
			...(result.createdAddress ? { createdAddress: result.createdAddress } : {}),
			...(result.createdAddresses ? { createdAddresses: result.createdAddresses } : {}),
		},
		method: 'tevm_call',
		...(request.id === undefined ? {} : { id: request.id }),
	}

	return out
}