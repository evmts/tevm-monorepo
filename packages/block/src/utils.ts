import { EthjsAddress, keccak256, toType as originalToType, toBytes } from '@tevm/utils'

/**
 * Creates a zero-filled buffer of the specified length
 * @param {number} length - The length of the buffer
 * @returns {Uint8Array} A zero-filled Uint8Array
 */
export function zeros(length: number): Uint8Array {
	return new Uint8Array(length)
}

/**
 * Creates the zero address
 * @returns {EthjsAddress} The zero address
 */
export function createZeroAddress() {
	return new EthjsAddress(new Uint8Array(20))
}

/**
 * Creates an address from a public key
 * @param {Uint8Array} publicKey - The public key
 * @returns {EthjsAddress} The address
 */
export function createAddressFromPublicKey(publicKey: Uint8Array): EthjsAddress {
	// Create address from public key by taking the last 20 bytes of keccak256 hash
	const addressBytes = keccak256(publicKey, 'bytes').slice(-20)
	return new EthjsAddress(addressBytes)
}

/**
 * Wrapper for toType that handles string inputs properly
 * @param {any} input - The input value
 * @param {number} outputType - The output type
 * @returns {any} The converted value
 */
export function safeToType(input: any, outputType: number): any {
	if (input === null || input === undefined) {
		return undefined
	}

	// Handle string inputs by converting to bytes first
	if (typeof input === 'string' && outputType === 2) {
		const bytes = toBytes(input)
		return originalToType(bytes, outputType as any)
	}

	return originalToType(input, outputType as any)
}

/**
 * Gets the v value from a signature
 * @param {any} signature - The signature object
 * @returns {bigint} The v value
 */
export function getSignatureV(signature: any): bigint {
	// Handle both old format (v) and new format (recovery)
	if (signature.v !== undefined) {
		return BigInt(signature.v)
	}
	if (signature.recovery !== undefined) {
		return BigInt(signature.recovery) + 27n
	}
	throw new Error('Invalid signature format: missing v or recovery')
}
