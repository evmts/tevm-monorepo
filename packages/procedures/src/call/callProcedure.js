import { callHandler } from '@tevm/actions'
import { hexToBigInt, numberToHex } from '@tevm/utils'
import { parseBlockTag } from '../utils/parseBlockTag.js'

/**
 * Creates a Call JSON-RPC Procedure for handling call requests with Ethereumjs EVM
 * @param {import('@tevm/node').TevmNode} client
 * @returns {import('./CallJsonRpcProcedure.js').CallJsonRpcProcedure}
 */
export const callProcedure = (client) => async (request) => {
	const { errors = [], ...result } = await callHandler(client)({
		throwOnFail: false,
		...(request.params[1]
			? {
					stateOverrideSet: Object.fromEntries(
						Object.entries(request.params[1]).map(([address, state]) => [
							address,
							{
								...(state.code ? { code: state.code } : {}),
								...(state.balance ? { balance: hexToBigInt(state.balance) } : {}),
								...(state.nonce ? { nonce: hexToBigInt(state.nonce) } : {}),
								...(state.state ? { state: state.state } : {}),
								...(state.stateDiff ? { stateDiff: state.stateDiff } : {}),
							},
						]),
					),
				}
			: {}),
		...(request.params[2]
			? {
					blockOverrideSet: {
						...(request.params[2].blobBaseFee ? { blobBaseFee: hexToBigInt(request.params[2].blobBaseFee) } : {}),
						...(request.params[2].baseFee ? { baseFee: hexToBigInt(request.params[2].baseFee) } : {}),
						...(request.params[2].gasLimit ? { gasLimit: hexToBigInt(request.params[2].gasLimit) } : {}),
						...(request.params[2].coinbase ? { coinbase: request.params[2].coinbase } : {}),
						...(request.params[2].time ? { time: hexToBigInt(request.params[2].time) } : {}),
						...(request.params[2].number ? { number: hexToBigInt(request.params[2].number) } : {}),
					},
				}
			: {}),
		...(request.params[0].code ? { code: request.params[0].code } : {}),
		...(request.params[0].blobVersionedHashes ? { blobVersionedHashes: request.params[0].blobVersionedHashes } : {}),
		...(request.params[0].caller ? { caller: request.params[0].caller } : {}),
		...(request.params[0].data ? { data: request.params[0].data } : {}),
		...(request.params[0].depth ? { depth: request.params[0].depth } : {}),
		...(request.params[0].gasPrice ? { gasPrice: hexToBigInt(request.params[0].gasPrice) } : {}),
		...(request.params[0].gas ? { gas: hexToBigInt(request.params[0].gas) } : {}),
		...(request.params[0].gasRefund ? { gasRefund: hexToBigInt(request.params[0].gasRefund) } : {}),
		...(request.params[0].origin ? { origin: request.params[0].origin } : {}),
		...(request.params[0].salt ? { salt: request.params[0].salt } : {}),
		...(request.params[0].selfdestruct ? { selfdestruct: new Set(request.params[0].selfdestruct) } : {}),
		...(request.params[0].skipBalance ? { skipBalance: request.params[0].skipBalance } : {}),
		...(request.params[0].to ? { to: request.params[0].to } : {}),
		...(request.params[0].value ? { value: hexToBigInt(request.params[0].value) } : {}),
		...(request.params[0].blockTag ? { blockTag: parseBlockTag(request.params[0].blockTag) } : {}),
		...(request.params[0].createTransaction ? { createTransaction: request.params[0].createTransaction } : {}),
		...(request.params[0].from ? { from: request.params[0].from } : {}),
		...(request.params[0].maxFeePerGas ? { maxFeePerGas: hexToBigInt(request.params[0].maxFeePerGas) } : {}),
		...(request.params[0].maxPriorityFeePerGas
			? { maxPriorityFeePerGas: hexToBigInt(request.params[0].maxPriorityFeePerGas) }
			: {}),
		// TODO add support for manually setting nonce
		// ...(request.params[0].nonce ? { nonce: hexToBigInt(request.params[0].nonce) } : {}),
	})
	if (errors.length > 0) {
		const error = /** @type {import('@tevm/actions').TevmCallError}*/ (errors[0])
		return {
			jsonrpc: '2.0',
			error: {
				code: error.code,
				message: error.message,
				data: {
					errors: errors.map(({ message }) => message),
				},
			},
			method: 'tevm_call',
			...(request.id === undefined ? {} : { id: request.id }),
		}
	}

	/**
	 * @type {Record<`0x${string}`, Array<import('@tevm/utils').Hex>> | undefined}
	 */
	const accessList =
		result.accessList !== undefined
			? Object.fromEntries(Object.entries(result.accessList).map(([key, value]) => [key, [...value]]))
			: undefined

	/**
	 * @param {bigint} value
	 * @returns {import('@tevm/utils').Hex}
	 */
	const toHex = (value) => /**@type {import('@tevm/utils').Hex}*/ (numberToHex(value))
	/**
	 * @type {import('./CallJsonRpcResponse.js').CallJsonRpcResponse}
	 */
	const out = {
		jsonrpc: '2.0',
		result: {
			executionGasUsed: toHex(result.executionGasUsed),
			rawData: result.rawData,
			...(result.selfdestruct ? { selfdestruct: [...result.selfdestruct] } : {}),
			...(result.gasRefund ? { gasRefund: toHex(result.gasRefund) } : {}),
			...(result.gas ? { gas: toHex(result.gas) } : {}),
			...(result.logs ? { logs: result.logs } : {}),
			...(result.txHash ? { txHash: result.txHash } : {}),
			...(result.blobGasUsed ? { blobGasUsed: toHex(result.blobGasUsed) } : {}),
			...(accessList !== undefined ? { accessList } : {}),
			...(result.preimages ? { preimages: result.preimages } : {}),
			...(result.l1Fee ? { l1Fee: numberToHex(result.l1Fee) } : {}),
			...(result.l1BaseFee ? { l1BaseFee: numberToHex(result.l1BaseFee) } : {}),
			...(result.l1BlobFee ? { l1BlobFee: numberToHex(result.l1BlobFee) } : {}),
			...(result.l1GasUsed ? { l1GasUsed: numberToHex(result.l1GasUsed) } : {}),
			...(result.amountSpent ? { amountSpent: numberToHex(result.amountSpent) } : {}),
			...(result.baseFee ? { baseFee: numberToHex(result.baseFee) } : {}),
			...(result.totalGasSpent ? { totalGasSpent: numberToHex(result.totalGasSpent) } : {}),
			...(result.trace
				? {
						trace: {
							...result.trace,
							gas: toHex(result.trace.gas),
							structLogs: result.trace.structLogs.map((log) => ({
								...log,
								gas: toHex(log.gas),
								gasCost: toHex(log.gasCost),
								stack: [...log.stack],
							})),
						},
					}
				: {}),
			...(result.createdAddress ? { createdAddress: result.createdAddress } : {}),
			...(result.createdAddresses ? { createdAddresses: [...result.createdAddresses] } : {}),
		},
		method: 'tevm_call',
		...(request.id === undefined ? {} : { id: request.id }),
	}

	return out
}
