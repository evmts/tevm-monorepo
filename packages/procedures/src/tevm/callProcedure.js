import { callHandler } from '@tevm/actions'
import { hexToBigInt, numberToHex } from '@tevm/utils'

/**
 * Creates a Call JSON-RPC Procedure for handling call requests with Ethereumjs EVM
 * @param {import('@tevm/base-client').BaseClient} client
 * @returns {import('@tevm/procedures-types').CallJsonRpcProcedure}
 */
export const callProcedure = (client) => async (request) => {
	const { errors = [], ...result } = await callHandler(client)({
		throwOnFail: false,
		...(request.params[0].deployedBytecode
			? { deployedBytecode: request.params[0].deployedBytecode }
			: {}),
		...(request.params[0].blobVersionedHashes
			? { blobVersionedHashes: request.params[0].blobVersionedHashes }
			: {}),
		...(request.params[0].caller ? { caller: request.params[0].caller } : {}),
		...(request.params[0].data ? { data: request.params[0].data } : {}),
		...(request.params[0].depth ? { depth: request.params[0].depth } : {}),
		...(request.params[0].gasPrice
			? { gasPrice: hexToBigInt(request.params[0].gasPrice) }
			: {}),
		...(request.params[0].gas
			? { gas: hexToBigInt(request.params[0].gas) }
			: {}),
		...(request.params[0].gasRefund
			? { gasRefund: hexToBigInt(request.params[0].gasRefund) }
			: {}),
		...(request.params[0].origin ? { origin: request.params[0].origin } : {}),
		...(request.params[0].salt ? { salt: request.params[0].salt } : {}),
		...(request.params[0].selfdestruct
			? { selfdestruct: new Set(request.params[0].selfdestruct) }
			: {}),
		...(request.params[0].skipBalance
			? { skipBalance: request.params[0].skipBalance }
			: {}),
		...(request.params[0].to ? { to: request.params[0].to } : {}),
		...(request.params[0].value
			? { value: hexToBigInt(request.params[0].value) }
			: {}),
		...(request.params[0].blockTag
			? { blockTag: request.params[0].blockTag }
			: {}),
		...(request.params[0].createTransaction
			? { createTransaction: request.params[0].createTransaction }
			: {}),
	})
	if (errors.length > 0) {
		const error = /** @type {import('@tevm/errors').CallError}*/ (errors[0])
		return {
			jsonrpc: '2.0',
			error: {
				code: error._tag,
				message: error.message,
				data: {
					errors: errors.map(({ message }) => message),
				},
			},
			method: 'tevm_call',
			...(request.id === undefined ? {} : { id: request.id }),
		}
	}
	/**
	 * @param {bigint} value
	 * @returns {import('@tevm/utils').Hex}
	 */
	const toHex = (value) =>
		/**@type {import('@tevm/utils').Hex}*/ (numberToHex(value))
	return {
		jsonrpc: '2.0',
		result: {
			executionGasUsed: toHex(result.executionGasUsed),
			rawData: result.rawData,
			...(result.selfdestruct
				? { selfdestruct: [...result.selfdestruct] }
				: {}),
			...(result.gasRefund ? { gasRefund: toHex(result.gasRefund) } : {}),
			...(result.gas ? { gas: toHex(result.gas) } : {}),
			...(result.logs ? { logs: result.logs } : {}),
			...(result.blobGasUsed ? { blobGasUsed: toHex(result.blobGasUsed) } : {}),
			...(result.createdAddress
				? { createdAddress: result.createdAddress }
				: {}),
			...(result.createdAddresses
				? { createdAddresses: [...result.createdAddresses] }
				: {}),
		},
		method: 'tevm_call',
		...(request.id === undefined ? {} : { id: request.id }),
	}
}
