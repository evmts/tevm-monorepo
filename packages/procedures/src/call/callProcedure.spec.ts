import { setAccountHandler } from '@tevm/actions'
import { ERC20 } from '@tevm/contract'
import { type TevmNode, createTevmNode } from '@tevm/node'
import { encodeFunctionData, numberToHex, parseEther } from '@tevm/utils'
import { beforeEach, describe, expect, it } from 'vitest'
import type { CallJsonRpcRequest } from './CallJsonRpcRequest.js'
import { callProcedure } from './callProcedure.js'

let client: TevmNode

beforeEach(() => {
	client = createTevmNode()
})

describe('callProcedure', () => {
	it('should handle a basic call', async () => {
		const address = `0x${'69'.repeat(20)}` as const

		await setAccountHandler(client)({
			address,
			balance: 420n,
			nonce: 69n,
			deployedBytecode: ERC20.deployedBytecode,
			state: {
				'0x0': '0x01',
			},
		})

		const request: CallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'tevm_call',
			id: 1,
			params: [
				{
					to: address,
					data: encodeFunctionData(ERC20.read.name()),
					gas: numberToHex(21000n),
					gasPrice: numberToHex(1n),
				},
			],
		}

		const response = await callProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('tevm_call')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle a call with state override', async () => {
		const to = `0x${'69'.repeat(20)}` as const
		const from = `0x${'42'.repeat(20)}` as const
		const request: CallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'tevm_call',
			id: 1,
			params: [
				{
					to,
					from,
					value: numberToHex(parseEther('.9')),
				},
				{
					[from]: {
						balance: numberToHex(parseEther('1')),
					},
				},
			],
		}

		const response = await callProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('tevm_call')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it.todo('should handle a call with block override', async () => {})

	it('should handle errors from callHandler', async () => {
		const request: CallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'tevm_call',
			id: 1,
			params: [
				{
					to: `0x${'00'.repeat(20)}` as const, // Invalid address
					from: `0x${'42'.repeat(20)}` as const,
					data: '0x0',
					value: numberToHex(500n),
					gas: numberToHex(21000n),
					gasPrice: numberToHex(1n),
				},
			],
		}

		const response = await callProcedure(client)(request)
		expect(response.error).toBeDefined()
		expect(response.error).toMatchSnapshot()
		expect(response.method).toBe('tevm_call')
		expect(response.id).toBe(request.id as any)
	})
})
