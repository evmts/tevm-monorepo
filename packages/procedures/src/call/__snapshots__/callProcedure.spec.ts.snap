// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`callProcedure should handle a basic call 1`] = `
{
  "amountSpent": "0x296a6",
  "createdAddresses": [],
  "executionGasUsed": "0xc62",
  "gas": "0x1c964d6",
  "logs": [],
  "rawData": "0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000",
  "selfdestruct": [],
  "totalGasSpent": "0x5eaa",
}
`;

exports[`callProcedure should handle a call with state override 1`] = `
{
  "amountSpent": "0x23e38",
  "executionGasUsed": "0x0",
  "rawData": "0x",
  "totalGasSpent": "0x5208",
}
`;

exports[`callProcedure should handle errors from callHandler 1`] = `
{
  "code": -32000,
  "data": {
    "errors": [
      
"sender doesn't have enough funds to send tx. The upfront cost is: 210000500 and the sender's account (0x4242424242424242424242424242424242424242) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0x66575cf759d7c97e42f103bfc8de6f88f70ea97a785d0c1f558496abf977cae1 nonce=0 value=500 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
    ],
  },
  "message": 
"sender doesn't have enough funds to send tx. The upfront cost is: 210000500 and the sender's account (0x4242424242424242424242424242424242424242) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0x66575cf759d7c97e42f103bfc8de6f88f70ea97a785d0c1f558496abf977cae1 nonce=0 value=500 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
}
`;
